File: spec/poc/src/reader.ts
// ─── SDF Reader ───────────────────────────────────────────────────────────────
// Implements the canonical consumer flow from SDF_FORMAT.md Section 6.
// Steps: open ZIP → read meta → validate data → return structured result

import { unpackContainer } from './container.js';
import { validateData, validateMeta } from './validator.js';
import {
  SDFArchive,
  SDFError,
  SDF_ERRORS,
} from './types.js';

// ─── parseSDF ─────────────────────────────────────────────────────────────────

export async function parseSDF(buffer: Buffer): Promise<SDFArchive> {
  // Step 1 — Open ZIP archive (SDF_FORMAT.md Section 6, Step 1)
  // TODO: container-decision — currently uses ZIP via unpackContainer()
  const contents = await unpackContainer(buffer);

  // Step 2 — Read and validate meta.json (SDF_FORMAT.md Section 6, Step 2)
  let meta: unknown;
  try {
    meta = JSON.parse(contents.metaJson);
  } catch {
    throw new SDFError(
      SDF_ERRORS.INVALID_META,
      'meta.json is not valid JSON.',
      undefined,
      'meta.json',
    );
  }
  validateMeta(meta); // throws SDFError if invalid — meta is now SDFMeta

  // Step 3 — Validate data.json against schema.json (SDF_FORMAT.md Section 6, Step 3)
  let schema: Record<string, unknown>;
  let data: Record<string, unknown>;

  try {
    schema = JSON.parse(contents.schemaJson) as Record<string, unknown>;
  } catch {
    throw new SDFError(
      SDF_ERRORS.INVALID_SCHEMA,
      'schema.json is not valid JSON.',
      undefined,
      'schema.json',
    );
  }

  try {
    data = JSON.parse(contents.dataJson) as Record<string, unknown>;
  } catch {
    throw new SDFError(
      SDF_ERRORS.SCHEMA_MISMATCH,
      'data.json is not valid JSON.',
      undefined,
      'data.json',
    );
  }

  const validation = validateData(data, schema);
  if (!validation.valid) {
    throw new SDFError(
      SDF_ERRORS.SCHEMA_MISMATCH,
      'data.json failed validation against schema.json.',
      validation.errors,
      'data.json',
    );
  }

  // Step 4 — Return structured result (SDF_FORMAT.md Section 6, Step 4 & 5)
  return {
    meta,
    data,
    schema,
    pdfBytes: contents.visualPdf,
  };
}

// ─── extractJSON ──────────────────────────────────────────────────────────────
// Partial read mode (SDF_FORMAT.md Section 12.3):
// Extracts meta + data without reading visual.pdf.
// Useful for high-throughput server-side ingestion where the PDF is not needed.

export async function extractJSON(buffer: Buffer): Promise<Pick<SDFArchive, 'meta' | 'data' | 'schema'>> {
  const result = await parseSDF(buffer);
  return { meta: result.meta, data: result.data, schema: result.schema };
}

File: packages/sdf-python/sdf/reader.py
# Copyright (c) 2026 Yunus YILDIZ — SPDX-License-Identifier: BUSL-1.1
"""SDF reader — parse .sdf archive. SDF_FORMAT.md Section 6."""

import json
import zipfile
from io import BytesIO
from typing import Any, Union

from . import _constants
from .errors import SDFError, SDF_ERRORS
from .types import SDFParseResult, SDFMeta
from .validator import validate_meta, validate_schema_or_throw


def _has_path_traversal(path: str) -> bool:
    return ".." in path or "\\" in path or path.startswith("/")


def _unpack_container(buffer: bytes) -> tuple[bytes, str, str, str]:
    """Unpack ZIP, return (visual_pdf, data_json, schema_json, meta_json)."""
    try:
        zf = zipfile.ZipFile(BytesIO(buffer), "r")
    except zipfile.BadZipFile as e:
        raise SDFError(SDF_ERRORS["NOT_ZIP"], "File is not a valid ZIP archive.") from e

    names = zf.namelist()
    for path in names:
        if _has_path_traversal(path):
            raise SDFError(
                SDF_ERRORS["INVALID_ARCHIVE"],
                f'Archive contains an invalid file path: "{path}".',
                file=path,
            )

    allowed = set(_constants.REQUIRED_FILES) | {"signature.sig"}
    for path in names:
        if zf.getinfo(path).is_dir():
            continue
        name = path.split("/")[0] if "/" in path else path
        if not (name in allowed or name.startswith("vendor/")):
            raise SDFError(
                SDF_ERRORS["INVALID_ARCHIVE"],
                f'Unexpected file at archive root: "{path}".',
                file=path,
            )

    total_size = 0
    for path in names:
        if zf.getinfo(path).is_dir():
            continue
        size = zf.getinfo(path).file_size
        if size > _constants.MAX_FILE_SIZE_BYTES:
            raise SDFError(
                SDF_ERRORS["ARCHIVE_TOO_LARGE"],
                f'File "{path}" exceeds the 50 MB per-file limit.',
                file=path,
            )
        total_size += size
        if total_size > _constants.MAX_TOTAL_SIZE_BYTES:
            raise SDFError(
                SDF_ERRORS["ARCHIVE_TOO_LARGE"],
                "Archive total uncompressed size exceeds the 200 MB limit.",
            )

    for required in _constants.REQUIRED_FILES:
        if required not in names:
            raise SDFError(
                SDF_ERRORS["MISSING_FILE"],
                f'Required file "{required}" is absent from the archive.',
                file=required,
            )

    visual_pdf = zf.read("visual.pdf")
    data_json = zf.read("data.json").decode("utf-8")
    schema_json = zf.read("schema.json").decode("utf-8")
    meta_json = zf.read("meta.json").decode("utf-8")
    zf.close()
    return visual_pdf, data_json, schema_json, meta_json


def parse_sdf(buffer: Union[bytes, bytearray]) -> SDFParseResult:
    """
    Parse an SDF archive. Returns meta, data, schema, and pdf_bytes.
    Raises SDFError on invalid structure.
    """
    data = bytes(buffer)
    visual_pdf, data_json, schema_json, meta_json = _unpack_container(data)

    meta_raw = json.loads(meta_json)
    if not isinstance(meta_raw, dict):
        raise SDFError(
            SDF_ERRORS["INVALID_META"],
            "meta.json is not valid JSON.",
            file="meta.json",
        )
    meta = validate_meta(meta_raw)

    schema_raw: Any = json.loads(schema_json)
    if not isinstance(schema_raw, dict):
        raise SDFError(
            SDF_ERRORS["INVALID_SCHEMA"],
            "schema.json is not a valid JSON object.",
            file="schema.json",
        )

    data_raw: Any = json.loads(data_json)
    if not isinstance(data_raw, dict):
        raise SDFError(
            SDF_ERRORS["SCHEMA_MISMATCH"],
            "data.json is not a valid JSON object.",
            file="data.json",
        )
    validate_schema_or_throw(data_raw, schema_raw)

    return SDFParseResult(
        meta=meta,
        data=data_raw,
        schema=schema_raw,
        pdf_bytes=visual_pdf,
    )
